src/Content/ContentStorage.php

```diff
@@ -262,32 +267,21 @@ protected function ensureExistingVersion(
 	 * used for storage
 	 *
 	 * @param bool $force If set to `true`, the language code is not validated
-	 * @return string Language code
 	 */
 	protected function language(
 		string|null $languageCode = null,
-		bool $force = false
-	): string {
-		// in force mode, use the provided language code even in single-lang for
-		// compatibility with the previous behavior in `$model->contentFile()`
-		if ($force === true) {
```

Comment:
We use this internally for the conversion only. That's why I wasn't worried about backwards compatibility. We have to pass this force parameter down a lot of methods so far just to manipulate the filename. That seemed like a very rough and dirty approach to me. But you might be right that the conversion could actually break but we don't catch it in the unit tests so far. I haven't double-checked that. 

In that case, I would suggest to handle this differently. I remember that we talked about it at btconf. The Language object could have a hint if it needs the language code prefix or not. We pass the Language object all the way down already - which makes a lot of sense and the content file methods should react to that and not to that added force argument.

Reply:
Sorry, I had a look at this again and it's starting to break my brain once more. 

The main reason why I had to remove the force option here is obviously the ContentStorage::language method that has to return a Language object in order to pass it down to the ContentStorageHandler. We can't simply "force" an invalid language code anymore at this point. 

For some reason, this does not break any unit tests, which is kind of shit and I basically relied on that. 

The content storage class will be removed entirely in one of the next PRs. That's why we are in a state right now where we would need to find a temporary fix for this for this particularly PR while it will change later again. 

I absolutely hate those situations. We are obviously lacking a proper unit test for single-language to multi-language conversion. I guess if we had one of those, it would probably fail with this PR. But what do we do? 

1) ignore the issue for now and add a todo for it? 
2) write a unit test in this PR and then create a temporary fix somehow only to replace it later again? 

I'm really not sure how to proceed.

Reply:
Ok, I sat down and I think I have a proper solution without delaying it further. 

1. I created a dedicated LanguageConversionTest class to make sure that we properly test single-to-multi-to-single language conversion
2. I've introduced a new single option for the Language model together with a new Language::isSingle method.  This is now used in Language::single to return a Language object with that state. 
3. ContentStorage::convertLanguage will now either be called with Language::single as first argument to convert form single to multi language, or as second argument to convert from multi language to single language. 
4. The PlainTextContentStorageHandler uses the Language::isSingle method to reliably determine if it should add the language code or not. 

I feel like this is a really clean way to handle this chaos. Another option we might want to consider is to use a SingleLanguage class, which extends the Language class instead of the ::isSingle state. This would maybe be a bit more predictable. What would you prefer?

Reply:
The language objects instead of strings with force are really great and are indeed cleaning it all up in a very good way.

About `isSingle` vs. `SingleLanguage`: Both have their advantages. `SingleLanguage` would show up more clearly when debugging, but since the single language will go with the new always-multilang concept, I don't think it's worth it.

Reply:
Good point. Then I would say that this PR is ready from my perspective